### PR TITLE
Add sis-local, a package for emulating SiS in local development

### PR DESCRIPTION
**Pull request by sfc-gh-jcarroll**
First effort at this, details in the README

### CODE REVIEW

File: sis-local/example/README.md
Comment (sfc-gh-amiribel) [inline sis-local/example/README.md]:
```suggestion
1. Create a secrets.toml file in this repo, as described in the parent directory README
2. Create and activate a python 3.8 virtual environment
3. Run the following
```

File: sis-local/streamlit_in_snowflake/sis.py
Comment (sfc-gh-amiribel) [inline sis-local/streamlit_in_snowflake/sis.py:23]:
```suggestion
# TODO: st.markdown should also warn when using unsafe html
# TODO: warn for custom components
# TODO: warn for components API
# TODO: st.info for cache_data and cache_resource? Since they won't crash but just won't work
```

File: sis-local/README.md
Comment (sfc-gh-amiribel) [inline sis-local/README.md:7]: I would perhaps vote to focus on the Streamlit-in-Snowflake local experience here, and not care too much about Snowpark session setup just per separation of concern. If Snowpark session setup is an identified pain, then perhaps that's worth another directory / README / doc?

Comment (sfc-gh-amiribel) [reply sis-local/README.md:7]: Or perhaps do you **need** to Snowpark somewhere to understand if you're in SiS / local?

File: sis-local/README.md
Comment (sfc-gh-amiribel) [inline sis-local/README.md]: Why `in_snowpark`?  Did you mean `in_sis()`? If the intended usecase for this block of code is to check whether one develops locally or in SiS, I'd just use something like this pattern instead maybe to make it clearer?

```
if developing_in_sis():
    ...
elif developing_locally():
    ...
else:
    raise Exception(...)
```

File: sis-local/example/README.md
Comment (sfc-gh-amiribel) [inline sis-local/example/README.md]: For a seamless local SiS experience, I think we should advocate for an Anaconda environment actually, which constraints to Snowflake Anaconda channel. That will prevent people from installing Streamlit components or other libraries which are not supported in SiS. What do you think? 

Example `environment.yml` and [instructions](https://github.com/snowflakedb/sis-datascience/tree/main#conda)
```
name: metrics
channels:
- snowflake
dependencies:
- python=3.8.12
- streamlit=1.13.0
- snowflake-snowpark-python
- plotly=5.9.0
- prophet=1.0.1
- tabulate=0.8.10
```

File: sis-local/README.md
Comment (blackary) [reply sis-local/README.md]: I agree on renaming in_snowpark with in_sis, though I don't think you need a second test for developing_locally

File: sis-local/example/README.md
Comment (blackary) [reply sis-local/example/README.md]: It is pretty neat, though, that markdown previews will automatically fix the numbering for you :)

File: sis-local/example/README.md
Comment (blackary) [reply sis-local/example/README.md]: +1 to this, especially since it's easier to pin yourself to python 3.8 this way. Though I would advocate for pip in general, I think it's worth using conda for a more seamless transition between local and SiS

File: sis-local/setup.py
Comment (blackary) [inline sis-local/setup.py:1]: TBH, I hadn't considered actually putting this in this repo, but that seems like a nice easy solution

File: sis-local/example/README.md
Comment (blackary) [reply sis-local/example/README.md]: Ooh, except we don't want to have to push the streamlit_in_snowflake package to conda, so that probably won't work. Maybe just doing pip is better for this. We could add pip dependencies to the environment.yml, but I'm not sure what would happen if that environment.yml was deployed to SiS -- I think it would just ignore the ones it can't install, but it might error out. It's a bit tricky.

We could just have an environment.yml like above, and then recommend they pip install this package after installing that. That might be the best middle-ground.

File: sis-local/README.md
Comment (blackary) [reply sis-local/README.md:7]: Yes, I think it's the easiest way to distinguish between local development and being in SiS, and I think it's reasonable to package both of those together (though my tendency has been to try and isolate individual components with this repo).

File: sis-local/example/sis_app.py
Comment (blackary) [inline sis-local/example/sis_app.py]: I'd be a little concerned about this misleading people to think that LocalSnowparkConnection is part of streamlit in general. I wonder about separating them out a bit in the package, and doing:

```python
import streamlit_in_snowflake as st
from streamlit_in_snowflake.local_connection import get_active_session
session = get_active_session()
```

WDYT?

File: sis-local/example/README.md
Comment (arnaudmiribel) [reply sis-local/example/README.md:10]: You can actually specify a pip entry within the conda environment.yml! Under a `pip` entry. See [docs](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-file-manually)

```
# run: conda env create --file environment.yml
name: sis
channels:
  - snowflake
dependencies:
  - python=3.8.*
  - pip
  - pip:
    - streamlit-in-snowflake
```

Comment (arnaudmiribel) [reply sis-local/example/README.md:10]: Oh you mean the above environment.yml may not work in SiS. Good point. Then they should do it manually, right

Comment (arnaudmiribel) [reply sis-local/example/README.md]: And we prioritize to add streamlit-in-snowflake in the Anaconda channel 😇

File: sis-local/README.md
Comment (sfc-gh-jcarroll) [reply sis-local/README.md:7]: Yeah I think it makes sense to support local session setup here because 98% of apps today will need to do it, it's another point of friction. Over time I think this will get easier so it won't have to live in this package but for now I think it's really useful (saves copying another 40 lines of code into every app)

Comment (sfc-gh-jcarroll) [reply sis-local/README.md:7]: I reworked it from the feedback below so it's more clearly separated - LMK what you think

File: sis-local/README.md
Comment (sfc-gh-jcarroll) [reply sis-local/README.md]: I changed it to `running_in_sis()` agree with Zachary it only needs one test

File: sis-local/setup.py
Comment (sfc-gh-jcarroll) [reply sis-local/setup.py:1]: If this gets traction, I would like to put it in its own repo so we can do easier version tagging since that will become a thing. But figured this was a good place to start :)

File: sis-local/example/sis_app.py
Comment (sfc-gh-jcarroll) [reply sis-local/example/sis_app.py]: Yep this makes sense. WDYT about the new version? I didn't want to confuse the local method with the snowpark API since it does behave a little differently. Again I don't think this part will be needed long term.

```python
if running_in_sis():
    import streamlit as st
    session = get_active_session()
else:
    import streamlit_in_snowflake as st
    from streamlit_in_snowflake.local_session import get_local_session
    session = get_local_session()
```

File: sis-local/example/README.md
Comment (sfc-gh-jcarroll) [reply sis-local/example/README.md]: I added a minimal working example, LMK what you think! I don't really want to be too prescriptive for external users. If there are tweaks to this, could we do them as a follow-on PR?

File: sis-local/README.md
Comment (sfc-gh-amiribel) [inline sis-local/README.md]:
```suggestion
as it will transition easily to the SiS environment. However, this package is not yet available in the [Snowflake Anaconda channel](https://repo.anaconda.com/pkgs/snowflake/),
```

File: sis-local/example/README.md
Comment (sfc-gh-amiribel) [inline sis-local/example/README.md]: Isn't this deprecated after we recommend using Conda? Or there should be a 
```
conda activate my_env
```
before that?

Comment (sfc-gh-amiribel) [reply sis-local/example/README.md]: (Oh sorry it just looks like this README and the main one diverged a little bit)

Comment (sfc-gh-jcarroll) [reply sis-local/example/README.md]: I wanted to put the minimal example here and the full conda example below, because someone CAN use this without conda and it may be overwhelming if they aren't familiar with it and just want to build a simple app.

The bullet point just above says "you need to create a virtual environment, we recommend doing it with conda".

Comment (sfc-gh-jcarroll) [reply sis-local/example/README.md]: I could also remove the requirements but still have `pip install <link to repo>` if the concern is about having both a requirements file and an environment file... Maybe that would be better. WDYT @blackary ?

Comment (sfc-gh-jcarroll) [inline sis-local/example/README.md]:
```suggestion
pip install "streamlit_in_snowflake @ git+https://github.com/sfc-gh-jcarroll/sis-tricks.git@sis-local#subdirectory=sis-local"
```

File: sis-local/README.md
Comment (blackary) [inline sis-local/README.md:30]: This would be a useful use-case of the walrus operator, but that's probably not worth using since
1. It's not very familiar
2. get_active_session() is probably fast to return

But, just for fun, I tried rewriting it, and it seemed reasonable. Mostly I just wanted an excuse to try out how the walrus operator feels, because I've rarely ever used it :) 

```python
from __future__ import annotations

from snowflake.snowpark.context import get_active_session
from snowflake.snowpark import Session
from snowflake.snowpark.exceptions import SnowparkSessionException

def get_sis_session() -> Session | None:
    try:
        return get_active_session()
    except SnowparkSessionException:
        return None

if session := get_sis_session():
    import streamlit as st
else:
    import streamlit_in_snowflake as st
    # This part is optional, and requires secrets.toml setup
    from streamlit_in_snowflake.local_session import get_local_session
    session = get_local_session()
```

File: sis-local/example/README.md
Comment (blackary) [reply sis-local/example/README.md]: Yeah, I think installing from this repo makes sense, so they don't have to do any manual copying and pasting of any extra files. I like that.

**Review by blackary — CHANGES_REQUESTED** [review]
I think switching to pip installing from the repo makes sense, and everything else looks good to me. Thanks for all your work on this @sfc-gh-jcarroll!

**Comment by sfc-gh-jcarroll** [comment]
OK, I updated the READMEs to reference the URL after this is merged. I _think_ it's ready to merge now? I don't have write access, want to push the button @blackary?

File: sis-local/example/README.md
Comment (blackary) [inline sis-local/example/README.md]:
```suggestion
pip install -r "streamlit_in_snowflake @ git+https://github.com/blackary/sis-tricks.git#subdirectory=sis-local"
```

Shouldn't this one also be the same?

**Review by blackary — APPROVED** [review]
:shipit:

File: sis-local/README.md
Comment (sfc-gh-jcarroll) [reply sis-local/README.md:30]: Neat! I wouldn't have thought to use this either.

For the sake of tribal knowledge - I _believe_ the way that get_active_session() is written today, it's quite efficient to call multiple times since it constructs the session prior to calls (somehow) https://github.com/snowflakedb/snowpark-python/blob/main/src/snowflake/snowpark/session.py#L144-L151

**Pull request merged**
